Beeper/BarcelonaMautrixIPC/Runtime/UnixSocketMautrixIPCChannel.swift

```diff
@@ -0,0 +1,83 @@
+import Foundation
+import Socket
+import Swog
+
+fileprivate let log = Logger(category: "UnixSocketMautrixIPCChannel")
+
+public class UnixSocketMautrixIPCChannel: MautrixIPCInputChannel, MautrixIPCOutputChannel {
+    private let socket: Socket
+    private var readThread: Thread? = nil
+    private var readRunLoop: CFRunLoop? = nil
+    private var readCallback: ((Data) -> ())? = nil
+    
+    public init(_ socketPath: String) {
+        do {
+            try socket = Socket.create(family:.unix)
+            try socket.connect(to: socketPath)
+            log.info("Connected to unix socket \(socketPath)")
+        } catch let error {
+            fatalError("Failed to connect unix socket \(error)")
+        }
+    }
+    
+    private func setupReadThread() {
+        readThread = Thread {
+            self.readRunLoop = CFRunLoopGetCurrent()
+            
+            RunLoop.current.schedule {
+                do {
+                    var shouldKeepRunning = true
+                    
+                    repeat {
+                        log.info("Waiting for data")
+                        var readData = Data(capacity: 4096)
+                        let bytesRead = try self.socket.read(into: &readData)
```

Comment:
This feels like a weird interface (passing a buffer to read into) and doesn't feel very swift-idiomatic. Also, could this set capacity for the buffer cause issues by potentially passing incomplete amounts of data to the `sharedBarcelonaStream`?

 I was thinking for a second of commenting that maybe you should use the `self.socket.readString()` method instead, but it seems that the rest of the pipeline expects data, so yeah, reading into this buffer is probably the move? What are your thoughts? Do we only ever receive utf8 text through this socket, and could thus safely convert everything to passing `String`s, or maybe we could read a string from the socket, then convert it to data and send it through the pipeline?

Also @tulir do you know if there's ever any non-utf8 data sent between barcelona <-> mautrix-imessage?

Reply:
It should be all JSON which means only text

Reply:
Ok, good to know.

Also, I read more of the BlueSocket documentation, and it looks like `readString()` only reads the data currently available in the socket and returns it as a string, so we would never get incomplete data (I think?), but I think that means that if there's no data currently available, we just get a nil string and have to keep looping and polling the socket until there's data available (which, now that I think of it, would maybe be the case with reading data into a buffer as well?). I don't think there's a way to get data, blocking until data appears, even though that's what seems would be ideal here.

With this in mind @bradtgmurray wouldn't this mean that if we ever happen to read from the socket when there's no extra data currently available, it'll break out of the loop and stop communicating with mautrix-imessage? Or am I misunderstanding something?

Reply:
It sounds to me like it'll break out as well. AFAIK sockets need a separate check with a file descriptor for whether they are ready for read ops.